src/storage/redis.rs

```diff
@@ -0,0 +1,400 @@
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct RedisStorage {
+    prefix: String,
+    pool: Pool<RedisConnectionManager>,
+    ttl: u64,
+    max_commands: isize,
+    lru_cache: Arc<Mutex<LruCache<String, Message>>>,
+}
+
+pub struct RedisStorageBuilder {
+    pool: Pool<RedisConnectionManager>,
+    prefix: String,
+    ttl: Duration,
+    cache_size: usize,
+    max_commands: isize,
+}
+
+impl RedisStorageBuilder {
+    pub fn new(pool: Pool<RedisConnectionManager>) -> RedisStorageBuilder {
+        RedisStorageBuilder {
+            pool: pool,
+            prefix: String::from("msgbus"),
+            ttl: Duration::from_secs(20),
+            max_commands: 500,
+            cache_size: 500,
+        }
+    }
+
+    pub fn prefix<S: Into<String>>(mut self, prefix: S) -> Self {
+        self.prefix = prefix.into();
+        self
+    }
+
+    pub fn ttl(mut self, ttl: Duration) -> Self {
+        self.ttl = ttl;
+        self
+    }
+
+    pub fn max_commands(mut self, max_commands: isize) -> Self {
+        self.max_commands = max_commands;
+        self
+    }
+
+    pub fn cache_size(mut self, cache_size: usize) -> Self {
+        self.cache_size = cache_size;
+        self
+    }
+
+    pub fn build(&self) -> RedisStorage {
+        RedisStorage::new(
+            &self.prefix,
+            self.pool.clone(),
+            self.ttl,
+            self.max_commands,
+            self.cache_size,
+        )
+    }
+}
+
+impl RedisStorage {
+    pub fn new<S: Into<String>>(
+        prefix: S,
+        pool: Pool<RedisConnectionManager>,
+        ttl: Duration,
+        max_commands: isize,
+        cache_size: usize,
+    ) -> Self {
+        let cache = Arc::new(Mutex::new(LruCache::new(cache_size)));
+
+        Self {
+            prefix: prefix.into(),
+            pool: pool,
+            ttl: ttl.as_secs(),
+            max_commands: max_commands,
+            lru_cache: cache,
+        }
+    }
+
+    pub async fn get_connection(&self) -> Result<PooledConnection<'_, RedisConnectionManager>> {
+        let conn = self
+            .pool
+            .get()
+            .await
+            .context("unable to retrieve a redis connection from the pool")?;
+
+        Ok(conn)
+    }
+
+    fn prefixed(&self, queue: Queue) -> String {
+        format!("{}.{}", self.prefix, queue)
+    }
+
+    async fn get_cached(&self, id: &str) -> Option<Message> {
+        self.lru_cache.lock().await.pop(id)
+    }
+
+    async fn cache(&self, id: &str, msg: &Message) -> Option<Message> {
+        let mut cache = self.lru_cache.lock().await;
+        cache.put(id.to_owned(), msg.clone())
+    }
+
+    pub fn builder(pool: Pool<RedisConnectionManager>) -> RedisStorageBuilder {
+        RedisStorageBuilder::new(pool)
+    }
+}
+
+struct ForwardedMessage {
+    pub id: String,
+    pub destination: u32,
+}
+
+impl ForwardedMessage {
+    pub fn from_str_pair<S: Into<String>>(pair: S) -> Result<Self> {
```

Comment:
There are traits in the std library to implement this. Also why the input has to be `Into<String>` something like `AsRef<str>` is fine because u don't need to own the input, instead you just need to parse it.

This instead need implement the `std::str::FromStr` trait. then you can simply do `parse()` on any string.